test/testdata/resolver/abstract_validation.rb

```diff
@@ -0,0 +1,158 @@
+  def bar; end
+end
+
+
+# it raises an error when defining an abstract class method on a
+# module
+module M1
+  extend T::Sig
+  extend T::Helpers
+  abstract!
+  sig {abstract.returns(Object)}
+  def self.foo; end # error: Static methods in a module cannot be abstract
+end
+
+# it fails if a concrete module doesn't implement abstract methods
+  module M2
+# ^^^^^^^^^ error: Missing definition for abstract method `AbstractMixin#bar`
+# ^^^^^^^^^ error: Missing definition for abstract method `AbstractMixin#foo`
+  extend T::Helpers
+  include AbstractMixin
+end
+
+# it fails if a class method is unimplemented
+class C3 < AbstractClass
+    # ^^ error: Missing definition for abstract method `AbstractClass.foo`
+  extend T::Sig
+  extend T::Helpers
+  sig { implementation.returns(Object) }
+  def bar; end
+end
+
+# it fails if an instance method is unimplemented
+  class C4 < AbstractClass
+# ^^^^^^^^^^^^^^^^^^^^^^^^ error: Missing definition for abstract method `AbstractClass#bar`
+  extend T::Sig
+  extend T::Helpers
+  sig {implementation.returns(Object)}
+  def self.foo; end
+end
+
+  # it handles splats and kwargs
+class SplatParent
+  extend T::Sig
+  extend T::Helpers
+  abstract!
+  sig { abstract.params(args: Integer, opts: Integer).void }
+  def foo(*args, **opts); end
+end
+
+class SplatChild1 < SplatParent
+  def foo(*args); end # error: Implementation of abstract method `SplatParent#foo` must accept **`opts`
+end
+
+class SplatChild2 < SplatParent
+  def foo(**opts); end # error: Implementation of abstract method `SplatParent#foo` must accept *`args`
+end
+
+
+module NoSigInInterface
+  extend T::Sig
+  extend T::Helpers
+  interface!
+
+  sig {abstract.returns(Object)}
+  def good; end
+
+  def bad; end # error: All methods in an interface must be declared abstract
+end
+
+
+module NonAbstractSigInInterface
+  extend T::Sig
+  extend T::Helpers
+  interface!
+
+  sig {abstract.returns(Object)}
+  def good; end
+
+  sig { returns(Object) }
+  def bad; end # error: All methods in an interface must be declared abstract
+end
+
+module UntypedMixin
+  def bad; end
+end
+module PrivateMethodInInterface
+  extend T::Sig
+  extend T::Helpers
+  interface!
+  include UntypedMixin # XXXXX
+
+  sig {abstract.returns(Object)}
+  def good; end
+end
+
+module PrivateMethodInInterface
+  extend T::Sig
+  extend T::Helpers
+  interface!
+
+  sig {abstract.returns(Object)}
+  private def bad; end # error: Interface method `PrivateMethodInInterface#bad` cannot be private
+end
+
+module ProtectedMethodInInterface
+  extend T::Sig
+  extend T::Helpers
+  interface!
+
+  sig {abstract.returns(Object)}
+  protected def bad; end # error: Interface method `ProtectedMethodInInterface#bad` cannot be protected
+end
+
+module GoodInterface
+  extend T::Sig
+  extend T::Helpers
+  interface!
+
+  sig {abstract.void}
+  def foo; end
+end
+
+module BadTypedImpl
+  extend T::Sig
+  extend T::Helpers
+
+  sig {implementation.returns(Integer)} # XXXXX
```

Comment:
what is this for?

Reply:
This is an error that's not yet checked statically, because it involves checking the types of the abstract/implementation definitions. I have that implemented on another branch.

Most of this file was copy/pasted from the runtime tests and then reworked to be a valid test—these snippets with `XXXX` are things that were present in those tests but don't get checked here yet. I'll replace them with more informative TODOs.